File: java/revenj-core/src/main/java/org/revenj/RevenjPermissionManager.java
package org.revenj;

import org.revenj.patterns.*;
import org.revenj.security.PermissionManager;
import org.revenj.security.GlobalPermission;
import org.revenj.security.RolePermission;
import org.revenj.security.UserPrincipal;
import rx.Observable;
import rx.Subscription;

import java.io.Closeable;
import java.security.Principal;
import java.util.*;
import java.util.concurrent.Callable;
import java.util.stream.Collectors;
import java.util.stream.Stream;

final class RevenjPermissionManager implements PermissionManager, Closeable {

	private final Callable<Optional<SearchableRepository<GlobalPermission>>> globalRepository;
	private final Callable<Optional<SearchableRepository<RolePermission>>> rolesRepository;
	private final boolean defaultPermissions;

	private final Subscription globalSubscription;
	private final Subscription roleSubscription;

	private Map<String, Boolean> globalPermissions = new HashMap<>();
	private Map<String, List<Pair>> rolePermissions = new HashMap<>();

	private Map<String, Boolean> cache = new HashMap<>();
	private final Map<Class<?>, List<Filter>> registeredFilters = new HashMap<>();

	private final class Pair {
		public final String name;
		public final boolean isAllowed;

		public Pair(String name, boolean isAllowed) {
			this.name = name;
			this.isAllowed = isAllowed;
		}
	}

	private final class Filter<T> {
		public final Specification<T> specification;
		public final String role;
		public final boolean inverse;

		public Filter(Specification<T> specification, String role, boolean inverse) {
			this.specification = specification;
			this.role = role;
			this.inverse = inverse;
		}
	}

	private boolean permissionsChanged = true;

	public RevenjPermissionManager(ServiceLocator locator) {
		this(locator.resolve(Properties.class),
				new Generic<Observable<Callable<GlobalPermission>>>() {
				}.resolve(locator),
				new Generic<Observable<Callable<RolePermission>>>() {
				}.resolve(locator),
				new Generic<Callable<Optional<SearchableRepository<GlobalPermission>>>>() {
				}.resolve(locator),
				new Generic<Callable<Optional<SearchableRepository<RolePermission>>>>() {
				}.resolve(locator));
	}

	public RevenjPermissionManager(
			Properties properties,
			Observable<Callable<GlobalPermission>> globalChanges,
			Observable<Callable<RolePermission>> roleChanges,
			Callable<Optional<SearchableRepository<GlobalPermission>>> globalRepository,
			Callable<Optional<SearchableRepository<RolePermission>>> rolesRepository) {
		String permissions = properties.getProperty("revenj.permissions");
		if (permissions != null && permissions.length() > 0) {
			if (!permissions.equalsIgnoreCase("open") && !permissions.equalsIgnoreCase("closed")) {
				throw new RuntimeException("Invalid revenj.permission settings found: '" + permissions + "'.\n"
						+ "Allowed values are open and closed");
			}
		}
		defaultPermissions = permissions == null || "open".equals(permissions);
		globalSubscription = globalChanges.subscribe(c -> permissionsChanged = true);
		roleSubscription = roleChanges.subscribe(c -> permissionsChanged = true);
		this.globalRepository = globalRepository;
		this.rolesRepository = rolesRepository;
	}

	private void checkPermissions() {
		if (!permissionsChanged) {
			return;
		}
		Optional<SearchableRepository<GlobalPermission>> global;
		Optional<SearchableRepository<RolePermission>> roles;
		try {
			global = globalRepository.call();
			roles = rolesRepository.call();
		} catch (Exception ignore) {
			global = Optional.empty();
			roles = Optional.empty();
		}
		if (global.isPresent()) {
			globalPermissions =
					global.get().search().stream().collect(
							Collectors.toMap(GlobalPermission::getName, GlobalPermission::getIsAllowed));
		}
		if (roles.isPresent()) {
			rolePermissions =
					roles.get().search().stream().collect(
							Collectors.groupingBy(
									RolePermission::getName,
									Collectors.mapping(it -> new Pair(it.getRoleID(), it.getIsAllowed()), Collectors.toList())));
		}
		cache = new HashMap<>();
		permissionsChanged = false;
	}

	private boolean checkOpen(String[] parts, int len) {
		if (len < 0) {
			return defaultPermissions;
		}
		String name = String.join(".", Arrays.copyOf(parts, len));
		Boolean found = globalPermissions.get(name);
		return found != null ? found : checkOpen(parts, len - 1);
	}

	private boolean implies(Principal principal, String role) {
		return principal instanceof UserPrincipal
				? ((UserPrincipal) principal).implies(role)
				: role.equals(principal.getName());
	}

	@Override
	public boolean canAccess(String identifier, Principal user) {
		checkPermissions();
		String target = identifier != null ? identifier : "";
		String id = user != null ? user.getName() + ":" + target : target;
		Boolean exists = cache.get(id);
		if (exists != null) {
			return exists;
		}
		String[] parts = target.split("\\.");
		boolean isAllowed = checkOpen(parts, parts.length);
		if (user != null) {
			List<Pair> permissions;
			for (int i = parts.length; i >= 0; i--) {
				String subName = String.join(".", Arrays.copyOf(parts, i));
				permissions = rolePermissions.get(subName);
				if (permissions != null) {
					Optional<Pair> found = permissions.stream().filter(it -> implies(user, it.name)).findFirst();
					if (found.isPresent()) {
						isAllowed = found.get().isAllowed;
						break;
					}
				}
			}
		}
		Map<String, Boolean> newCache = new HashMap<>(cache);
		newCache.put(id, isAllowed);
		cache = newCache;
		return isAllowed;
	}

	@Override
	public <T, S extends T> Query<S> applyFilters(Class<T> manifest, Principal user, Query<S> data) {
		if (user == null) return data.filter(it -> defaultPermissions);
		List<Filter> registered = registeredFilters.get(manifest);
		if (registered != null) {
			Query<S> result = data;
			for (Filter r : registered) {
				if ((implies(user, r.role)) != r.inverse) {
					result = result.filter(r.specification);
				}
			}
			return result;
		}
		return data;
	}

	@Override
	public <T, S extends T> List<S> applyFilters(Class<T> manifest, Principal user, List<S> data) {
		if (user == null) return defaultPermissions ? data : Collections.EMPTY_LIST;
		List<Filter> registered = registeredFilters.get(manifest);
		if (registered != null) {
			Stream<S> result = data.stream();
			boolean filtered = false;
			for (Filter r : registered) {
				if (implies(user, r.role) != r.inverse) {
					result = result.filter(r.specification);
					filtered = true;
				}
			}
			return filtered ? result.collect(Collectors.toList()) : data;
		}
		return data;
	}

	@Override
	public <T> Closeable registerFilter(Class<T> manifest, Specification<T> filter, String role, boolean inverse) {
		List<Filter> registered = registeredFilters.get(manifest);
		if (registered == null) {
			registered = new ArrayList<>();
			registeredFilters.put(manifest, registered);
		}
		Filter item = new Filter(filter, role, inverse);
		List<Filter> reg = registered;
		reg.add(item);
		return () -> reg.remove(item);
	}

	public void close() {
		globalSubscription.unsubscribe();
		roleSubscription.unsubscribe();
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/RevenjPermissionManager.scala
package net.revenj

import java.security.Principal
import java.util.Properties

import monix.reactive.Observable
import net.revenj.patterns.{SearchableRepository, ServiceLocator}
import net.revenj.security.{GlobalPermission, PermissionManager, RolePermission, UserPrincipal}

private[revenj] class RevenjPermissionManager(
  properties: Properties,
  staticGlobalPermissions: Option[Seq[GlobalPermission]],
  staticRolePermissions: Option[Seq[RolePermission]]/*,
  globalChanges: Observable[Function0[GlobalPermission]],
  rolesChanges: Observable[Function0[RolePermission]],
  globalRepository: Function0[Option[SearchableRepository[GlobalPermission]]],
  rolesRepository: Function0[Option[SearchableRepository[RolePermission]]]*/
) extends PermissionManager {

  def this(properties: Properties, locator: ServiceLocator) {
    this(properties,
      locator.resolve[Option[Seq[GlobalPermission]]],
      locator.resolve[Option[Seq[RolePermission]]]/*,
      locator.resolve[Observable[Function0[GlobalPermission]]],
      locator.resolve[Observable[Function0[RolePermission]]],
      locator.resolve[Function0[Option[SearchableRepository[GlobalPermission]]]],
      locator.resolve[Function0[Option[SearchableRepository[RolePermission]]]]*/
    )
  }

  private val defaultPermissions = {
    val permissions = properties.getProperty("revenj.permissions")
    if (permissions != null && permissions.length > 0) {
      if (!permissions.equalsIgnoreCase("open") && !permissions.equalsIgnoreCase("closed")) throw new RuntimeException("Invalid revenj.permission settings found: '" + permissions + "'.\n" + "Allowed values are open and closed")
      "open" == permissions
    } else false
  }

  private var globalPermissions = staticGlobalPermissions.getOrElse(Nil).map(it => it.name -> it.isAllowed).toMap
  private var rolePermissions = staticRolePermissions.getOrElse(Nil).groupBy(_.name)

  private var cache = Map.empty[String, Boolean]
/*
  import monix.execution.Scheduler.Implicits.global

  globalChanges.doOnNext { _ =>
    globalRepository() match {
      case Some(rep) =>
        import scala.concurrent.ExecutionContext.Implicits.global
        rep.search().map { found =>
          globalPermissions = found.map(it => it.name -> it.isAllowed).toMap
        }
      case _ =>
    }
  }.subscribe()

  rolesChanges.doOnNext { _ =>
    rolesRepository() match {
      case Some(rep) =>
        import scala.concurrent.ExecutionContext.Implicits.global
        rep.search().map { found =>
          rolePermissions = found.groupBy(_.name)
        }
    }
  }.subscribe()
*/
  private def checkOpen(parts: Array[String], len: Int): Boolean = {
    if (len < 0) defaultPermissions
    else {
      val name = parts.take(len).mkString(".")
      globalPermissions.get(name) match {
        case Some(found) => found
        case _ => checkOpen(parts, len - 1)
      }
    }
  }

  private def implies(principal: Principal, role: String) = {
    principal match {
      case up: UserPrincipal => up.implies(role)
      case _ => principal.getName == role
    }
  }

  override def canAccess(identifier: String)(implicit user: Principal): Boolean = {
    val target = if (identifier != null) identifier else ""
    val id = if (user != null) user.getName + ":" + target else target
    cache.get(id) match {
      case Some(exists) => exists
      case _ =>
        val parts = target.split("\\.")
        var isAllowed = checkOpen(parts, parts.length)
        if (user != null) {
          var i = parts.length
          while (i >= 0) {
            val subName = parts.take(i).mkString(".")
            rolePermissions.get(subName) match {
              case Some(permissions) =>
                permissions.find(it => implies(user, it.name)) match {
                  case Some(found) =>
                    isAllowed = found.isAllowed
                    i = 0
                  case _ =>
                }
              case _ =>
            }
            i -= 1
          }
        }
        cache = cache + (id -> isAllowed)
        isAllowed
    }
  }
}